seqr/views/apis/saved_variant_api.py

```diff
@@ -0,0 +1,511 @@
+            lifted_over_chrom, lifted_over_pos = get_chrom_pos(saved_variant.lifted_over_xpos_start)
+        else:
+            coords_field = 'grch%s_coords' % lifted_over_genome_version
+            coords = variant_json.get('extras', {}).get(coords_field, '').split('-')
+            lifted_over_chrom = coords[0].lstrip('chr')
+            lifted_over_pos = coords[1] if len(coords) > 1 else ''
+        variant = {
+            'variantId': saved_variant.guid,
+            'xpos': saved_variant.xpos_start,
+            'ref': saved_variant.ref,
+            'alt': saved_variant.alt,
+            'chrom': chrom,
+            'pos': pos,
+            'genomeVersion': genome_version,
+            'liftedOverGenomeVersion': lifted_over_genome_version,
+            'liftedOverChrom': lifted_over_chrom,
+            'liftedOverPos': lifted_over_pos,
+            'familyGuid': saved_variant.family.guid,
+            'tags': _variant_tags(saved_variant),
+            'functionalData': _variant_functional_data(saved_variant),
+            'notes': _variant_notes(saved_variant),
+        }
+        if variant['tags'] or variant['notes']:
+            variant.update(_variant_details(variant_json, request.user))
+            variants[variant['variantId']] = variant
+
+    return create_json_response({'savedVariants': variants})
+
+
+@login_required(login_url=API_LOGIN_REQUIRED_URL)
+@csrf_exempt
+def create_variant_note_handler(request, variant_guid):
+    saved_variant = SavedVariant.objects.get(guid=variant_guid)
+    check_permissions(saved_variant.project, request.user, CAN_EDIT)
+
+    request_json = json.loads(request.body)
+    create_seqr_model(
+        VariantNote,
+        saved_variant=saved_variant,
+        note=request_json.get('note'),
+        submit_to_clinvar=request_json.get('submitToClinvar', False),
+        search_parameters=request_json.get('searchParameters'),
+        created_by=request.user,
+    )
+
+    return create_json_response({variant_guid: {'notes': _variant_notes(saved_variant)}})
+
+
+@login_required(login_url=API_LOGIN_REQUIRED_URL)
+@csrf_exempt
+def update_variant_note_handler(request, variant_guid, note_guid):
+    saved_variant = SavedVariant.objects.get(guid=variant_guid)
+    check_permissions(saved_variant.project, request.user, CAN_EDIT)
+    note = VariantNote.objects.get(guid=note_guid, saved_variant=saved_variant)
+
+    request_json = json.loads(request.body)
+    update_model_from_json(note, request_json, allow_unknown_keys=True)
+
+    return create_json_response({variant_guid: {'notes': _variant_notes(saved_variant)}})
+
+
+@login_required(login_url=API_LOGIN_REQUIRED_URL)
+@csrf_exempt
+def delete_variant_note_handler(request, variant_guid, note_guid):
+    saved_variant = SavedVariant.objects.get(guid=variant_guid)
+    check_permissions(saved_variant.project, request.user, CAN_EDIT)
+    note = VariantNote.objects.get(guid=note_guid, saved_variant=saved_variant)
+    delete_seqr_model(note)
```

Comment:
if the SavedVariant doesn't have any remaining notes or tags, it should also be deleted.
You've already taken care of most of the complexity of having SavedVariant be a separate model, but what do you think about having SavedVariant as an (abstract) base class of VariantTag, VariantNote, etc., rather than a separate table - similar to ModelWithGuid?

Reply:
1) Right now if you quickly remove a tag using the shortcut toggle and then add another one and there is a brief second where the saved variant has no tags but it is still visible to the user. If we deleted it in that second it would be entirely lost to the user which would be bad UX. Maybe we have a cleanup manage command that clears out old saved variants periodically?

2) That approach makes querying for all the saved variants really annoying since you get tons of duplicate variant data and then have to merge all the shared variant fields and transform the tag and note specific fields into into lists, whereas having the single variant model naturally keeps that data in the form we want it anyways. Additionally, the merge logic becomes gross: what happens when the saved variant json differs slightly between tags on the same variant? How do we keep that in sync? If we let them get out of sync you could get really inconsistent bugs where sometimes the correct variant data is showing and other times incorrect data is.

Reply:
agree - the way you have it is better